rsp_vision/dashboard/pages/helpers/calculations_for_plotting.py

```diff
@@ -0,0 +1,162 @@
+        appended to the original dataframe at the end.
+    """
+    mean_df = (
+        signal.groupby(["sf", "tf", "stimulus_frames"])
+        .agg({"signal": "mean"})
+        .reset_index()
+    )
+    mean_df["stimulus_repetition"] = "mean"
+    combined_df = pd.concat([signal, mean_df], ignore_index=True)
+
+    median_df = (
+        signal.groupby(["sf", "tf", "stimulus_frames"])
+        .agg({"signal": "median"})
+        .reset_index()
+    )
+    median_df["stimulus_repetition"] = "median"
+    combined_df = pd.concat([combined_df, median_df], ignore_index=True)
+
+    return combined_df
+
+
+def find_peak_coordinates(
+    fitted_gaussian_matrix: np.ndarray,
+    spatial_frequencies: np.ndarray,
+    temporal_frequencies: np.ndarray,
+    matrix_definition: int,
+) -> tuple:
+    """This method finds the peak coordinates of the fitted gaussian matrix.
+
+    Parameters
+    ----------
+    fitted_gaussian_matrix : np.ndarray
+        The fitted gaussian matrix obtained from the precalculated fits.
+    spatial_frequencies : np.ndarray
+        The spatial frequencies that are used in the experiment.
+    temporal_frequencies : np.ndarray
+        The temporal frequencies that are used in the experiment.
+    matrix_definition : int
+        The matrix definition used to generate the fitted_gaussian_matrix.
+
+    Returns
+    -------
+    tuple
+        The peak coordinates of the fitted gaussian matrix.
+    """
+    peak_indices = np.unravel_index(
+        np.argmax(fitted_gaussian_matrix), fitted_gaussian_matrix.shape
+    )
+
+    spatial_freq_linspace = np.linspace(
+        spatial_frequencies.min(),
+        spatial_frequencies.max(),
+        matrix_definition,
+    )
+    temporal_freq_linspace = np.linspace(
+        temporal_frequencies.min(),
+        temporal_frequencies.max(),
+        matrix_definition,
+    )
+
+    sf = spatial_freq_linspace[peak_indices[0]]
+    tf = temporal_freq_linspace[peak_indices[1]]
+    return tf, sf
+
+
+def call_get_gaussian_matrix_to_be_plotted(
+    n_roi: int,
+    fit_outputs: dict,
+    spatial_frequencies: np.ndarray,
+    temporal_frequencies: np.ndarray,
+    matrix_definition: int,
+) -> dict:
+    """This method is a wrapper for the get_gaussian_matrix_to_be_plotted
+    method that iterates over all the ROIs.
+
+    Parameters
+    ----------
+    n_roi : int
+        The number of ROIs.
+    fit_outputs : dict
+        The fit outputs obtained from the precalculated fits.
+    spatial_frequencies : np.ndarray
+        The spatial frequencies that are used in the experiment.
+    temporal_frequencies : np.ndarray
+        The temporal frequencies that are used in the experiment.
+    matrix_definition : int
+        The matrix definition used to generate the fitted_gaussian_matrix.
+
+    Returns
+    -------
+    dict
+        The fitted gaussian matrix obtained from the precalculated fits.
+    """
+    fitted_gaussian_matrix = {}
+
+    for roi_id in range(n_roi):
```

Comment:
This is cool syntax I don't think I'm aware of it, what is going on here?

Reply:
wait, what is the cool syntax here?